### DISCUSSION TITLE
Edit-and-continue changes async codegen; make it configurable

### DISCUSSION THREAD

**Discussion by jthorborg** [dotnet/csharplang#4432] [Language Ideas] [answered]
`async`/`await` and "task-like" friends is a really interesting system that actually allows to conjure a completely unmanaged stackless coroutine system, without ever involving the GC. This allows to build performant graph processing systems for games that do not suffer from recursion (and thereby stack overflow) for instance.

Unfortunately, the lowering of the feature changes codegen when running compiling under debug in an effort to support edit and continue - specifically the generated closure is a `class` under debug:
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.cs#L68

This makes it impossible to constrain generic arguments in methods in the builder pointed to by `AsyncMethodBuilder` to be `: unmanaged`.

Worse, lowering happens _after_ proposed source generation and compilation checks _before_ assembly is emitted, so the usual suspects like Cecil cannot be employed to solve this.

I'm working in a domain where edit and continue would never work anyway, so a reasonable compromise could be having a compiler flag for overriding whether EnC is enabled, looks like `CompilationOptions` was already encapsulating this perhaps with that in mind?

Thanks for any thoughts!

**Comment by bernd5** [comment, marked as answer]
This question seems to be belong to the roslyn repo - not csharplang.
Lowering is an implementation detail - it is not enforced by the language.

**Reply by jthorborg** [reply]
Absolutely. I seemed to have been redirected here, I started in the Roslyn repo.

**Reply by jthorborg** [reply]
Reposted in Roslyn: https://github.com/dotnet/roslyn/discussions/51189

This discussion can be closed now.
